feat: dance break channel

vzaharee opened the pull request:
<!--- Please fill out the following template. -->
<!--- Do not remove the template -->

### Description
<!--- Explain what your change adds/or fixes. -->
Adds a StepMania/DDR/In The Groove break channel.
<!--- If your change is a new channel, explain what it does and how it responds to different events. Please include a screenshot and/or (preferably) a video -->
Spawns a steady stream of randomly generated dance notes. When notes reach the top they are scored as "GOOD!" by default. Every donation will cause a note to score as "GREAT!", "PERFECT!", or "MARVELOUS" instead. If multiple donations occur in a row a combo counter will appear which will be reset the next time a note is scored "GOOD!". If donations come in faster than notes are being generated, then eighth or potentially sixteenth notes will be added in an attempt to keep up. A random song background is chosen each time the channel loads.

![](https://i.imgur.com/bLb1Idj.png)
![](https://i.imgur.com/eTbHny5.png)
![](https://i.imgur.com/EOk8EoM.png)

<!--- If your change adds any new dependencies or has any special requirements for being ran, such as a connection to an external service, please make a note of those additions/requirements and why they are necessary. -->

### Checklist:
<!--- Add an `x` to the boxes once you've confirmed them. -->
- [x] I have read and followed the requirements in the [**Contributing** document](https://github.com/GamesDoneQuick/gdq-break-channels/blob/main/CONTRIBUTING.md).
- [x] My code has been tested.
- [x] My commit title follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) formatting.
- [x] All the code is my own, or is code I have the rights to, and is being made available under the [Apache License Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).

<!-- If your change is for a new channel, uncomment and check the following -->
- [x] My channel contains no first-party Nintendo assets.
- [x] I understand that GamesDoneQuick may reject, make changes to, or choose not to show this channel on broadcast, and that discussion or merging does not guarantee the work will be used.

chrishanel reviewed (changes requested):
Hey, thanks for the submission!

The concept for this is strong (Dillon used it as his example for a potential channel idea after all), but I think the relationship between donations and this channel's behavior needs some adjustments. Primarily, the pace of donations required in order to experience a satisfying combo is extremely rare during an event, if not impossible. You're also increasing the step rate when donations come in, which works against keeping the combo going.

One thing I would suggest is changing it so that instead of combos being extended by individual donations, having it where every dollar counts as 1 combo step. In this situation I think the dynamic step rate works a lot better.

Two other notes:

1) I am not seeing a backdrop for the channel, it's just a black screen

2) Would you be interested in adding a UI element denoting a high combo score?

vzaharee commented:
Hi!

Yeah, I see what you're saying. I don't have a good sense for what the rates of donations or donation amounts are, so was struggling trying to figure out what would work here.

Maybe the square root of the donation amount (or some other formula) would be a good solution here? Just doing some back of the envelope math, with $1 = 1 combo, a $1000 donation alone at 120 BPM & sixteenth notes (8 notes/second) would last 125 seconds:

| $ | floor(sqrt($)) hits
| - | - |
| 1 | 1 |
| 5 | 2 |
| 10 | 3 |
| 50 | 7 |
| 100 | 10 |
| 500 | 22 |
| 1000 | 31 |
| 5000 | 70 |
| 10000 | 100 |

I think a high combo makes sense, and I'll try to figure out the black screen.

vzaharee commented:
I think tweaking the thresholds of when we move from every other beat, every beat, eighth notes, and sixteenth notes higher would help keep combos going while still providing some amount of dynamism/a catch up mechanic.

vzaharee commented:
@chrishanel I added a max combo element, fixed the background, raised the thresholds at which note speed increases occur, and used the square root of the donation amount for how much to extend the combo, if you wanted to give it another try. :slightly_smiling_face:

VodBox reviewed (changes requested):
Because the backgrounds are noisey, and some of them are pretty bright, the text in a lot of cases is not particularly legible, especially the hit messages as they start to fade out immediately. Would consider either making the backgrounds darker, adding some kind of text shadow, and definitely make the fade out wait a bit before starting.

Have you considered adding/using a different font or images for the different hit messages? I feel gdqpixel is not the best font for that kind of thing.

The donation text itself just being flat white in the corner doesn't really feel integrated or part of the channel as a whole and feels more like an after-thought. Is there a way it could be styled to match the aesthetics of a dance rhythm game, or make it feel integrated with the rest of the channel? Maybe treating it like a score indicator in a HUD?

It'd be nice if the background could have some motion to it, either with an animated texture (scanlines?) or maybe with it scaling/reacting to the combos? That said, care should be taken if doing that to not overdo the motion and be too distracting or attention grabbing.

vzaharee commented:
Those all sound pretty reasonable to me.

TBH the choice of gdqpixel was convenience (and I wasn't particularly happy with how the "B" in "COMBO" looked), and the rough edges in general a result of seeing a the Github channel yesterday and trying to get _something_ put together before hopping on a plane to Minneapolis tomorrow. :sweat_smile: 

It looks like the dance game community is pretty obsessive about fonts/theming though (I just learned the term for the hit messages is "judgement"), so I don't think it'll be too hard to find something with the right look and license.

I'll take a look at making these changes when I get back next week.
